src/Microsoft.Diagnostics.NETCore.Client/ReversedServer/ReversedDiagnosticsServer.cs

```diff
@@ -0,0 +1,164 @@
+        /// to the location specified by <paramref name="transportPath"/>.
+        /// </summary>
+        /// <param name="transportPath">
+        /// The path of the server endpoint.
+        /// On Windows, this can be a full pipe path or the name without the "\\.\pipe\" prefix.
+        /// On all other systems, this must be the full file path of the socket.
+        /// </param>
+        /// <param name="maxConnections">The maximum number of connections the server will support.</param>
+        public ReversedDiagnosticsServer(string transportPath, int maxConnections)
+        {
+            _transport = IpcServerTransport.Create(transportPath, maxConnections);
+        }
+
+        public void Dispose()
+        {
+            if (!_disposed)
+            {
+                _cancellation.Cancel();
+
+                IEnumerable<ServerIpcEndpoint> endpoints = _endpoints.Values;
+                _endpoints.Clear();
+
+                foreach (var endpoint in endpoints)
+                {
+                    endpoint.Dispose();
+                }
+
+                _transport.Dispose();
+
+                _cancellation.Dispose();
+
+                _disposed = true;
+            }
+        }
+
+        /// <summary>
+        /// Provides connection information when a new runtime instance connects to the server.
+        /// </summary>
+        /// <param name="token">The token to monitor for cancellation requests.</param>
+        /// <returns>A <see cref="ReversedDiagnosticsConnection"/> that contains information about the new runtime instance connection.</returns>
+        /// <remarks>
+        /// This will only provide connection information on the first time a runtime connects to the server. Subsequent
+        /// reconects will update the existing <see cref="ReversedDiagnosticsConnection"/> instance. If a connection is removed
+        /// using <see cref="RemoveConnection(Guid)"/> and the same runtime instance reconnects afte this call, then a
+        /// new <see cref="ReversedDiagnosticsConnection"/> will be produced.
+        /// </remarks>
+        public async Task<ReversedDiagnosticsConnection> AcceptAsync(CancellationToken token)
+        {
+            VerifyNotDisposed();
+
+            using CancellationTokenSource linkedSource = CancellationTokenSource.CreateLinkedTokenSource(token, _cancellation.Token);
+
+            ReversedDiagnosticsConnection newConnection = null;
+            do
+            {
+                Stream stream = null;
+                IpcAdvertise advertise = null;
+                try
+                {
+                    stream = await _transport.AcceptAsync(linkedSource.Token);
+                    advertise = IpcAdvertise.Parse(stream);
+                }
+                catch (Exception ex) when (!(ex is OperationCanceledException))
+                {
+                    // The advertise data could be incomplete if the runtime shuts down before completely writing
```

Comment:
What if the runtime writes part of the message and then hangs without exiting? I imagine the code here blocks indefinitely if that occurs.

As an error in a well-behaved client it seems unlikely but as a DOS attack from a malicious client it would be trivial.

Reply:
I've changed the parsing to be async and cancellable. The reversed server will cancel the parsing of the stream data after 3 seconds. I can add a test to check that if the client writes partial data and doesn't close the stream that it does not hang the server indefinitely.